File: Source/Engine/HyperRendering/HyperVulkan/src/HyperVulkan/Surface.cpp
/*
 * Copyright (c) 2020-2021, SkillerRaptor <skillerraptor@protonmail.com>
 *
 * SPDX-License-Identifier: MIT
 */

#include "HyperVulkan/Surface.hpp"

#include "HyperVulkan/Context.hpp"

#include <volk.h>
#include <GLFW/glfw3.h>

namespace HyperRendering::Vulkan
{
	Surface::Surface(Context& context)
		: m_context(context)
	{
	}

	auto Surface::initialize() -> HyperCore::Result<void, HyperCore::Errors::ConstructError>
	{
		if (glfwCreateWindowSurface(m_context.instance(), m_context.window().native_window(), nullptr, &m_surface) != VK_SUCCESS)
		{
			HyperCore::Logger::fatal("Failed to create vulkan window surface");
			return HyperCore::Errors::ConstructError::Incomplete;
		}
		
		HyperCore::Logger::debug("Vulkan window surface was created");
		
		return {};
	}

	auto Surface::terminate() -> void
	{
		vkDestroySurfaceKHR(m_context.instance(), m_surface, nullptr);
		HyperCore::Logger::debug("Vulkan window surface was destroyed");
	}
	
	auto Surface::surface() const -> VkSurfaceKHR
	{
		return m_surface;
	}
} // namespace HyperRendering::Vulkan

File: src/rendering/surface.rs
/*
 * Copyright (c) 2022-present, SkillerRaptor <skillerraptor@protonmail.com>
 *
 * SPDX-License-Identifier: MIT
 */

use crate::core::window::Window;
use crate::rendering::error::Error;
use crate::rendering::instance::Instance;

use ash::vk;
use log::debug;

pub struct Surface {
    pub surface: vk::SurfaceKHR,
    pub surface_loader: ash::extensions::khr::Surface,
}

impl Surface {
    pub fn new(
        window: &Window,
        entry: &ash::Entry,
        instance: &std::rc::Rc<Instance>,
    ) -> Result<Self, Error> {
        let surface_loader = ash::extensions::khr::Surface::new(entry, &instance.instance);
        let surface = unsafe {
            ash_window::create_surface(entry, &instance.instance, &window.native_window, None)?
        };

        debug!("Created vulkan surface");
        Ok(Self {
            surface,
            surface_loader,
        })
    }
}

impl Drop for Surface {
    fn drop(&mut self) {
        unsafe {
            self.surface_loader.destroy_surface(self.surface, None);
        }
    }
}
